File: graphs/number-of-islands/sol.py
# https://leetcode.com/problems/number-of-islands/

from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        visited = set()
        m, n = len(grid), len(grid[0])

        def dfs(i, j):
            if (
                i not in range(m)
                or j not in range(n)
                or grid[i][j] == "0"
                or (i, j) in visited
            ):
                return
            else:
                visited.add((i, j))
                dfs(i, j + 1)
                dfs(i + 1, j)
                dfs(i, j - 1)
                dfs(i - 1, j)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and (i, j) not in visited:
                    ans += 1
                    dfs(i, j)

        return ans

File: graphs/number-of-islands/sol.cpp
// https://leetcode.com/problems/number-of-islands/

#include <vector>

using std::vector;

class Solution
{
public:
  int numIslands(vector<vector<char>> &grid)
  {
    int ans = 0;
    int m = grid.size();
    int n = grid.at(0).size();

    for (int i = 0; i < m; i++)
    {
      for (int j = 0; j < n; j++)
      {
        if (grid.at(i).at(j) == '1')
        {
          dfs(grid, i, j, m, n);
          ans++;
        }
      }
    }

    return ans;
  }

private:
  void dfs(vector<vector<char>> &grid, int i, int j, int m, int n)
  {
    if (i < 0 || i >= m || j < 0 || j >= n || grid.at(i).at(j) == '0')
    {
      return;
    }
    else
    {
      grid.at(i).at(j) = '0'; // mark visited
      dfs(grid, i, j + 1, m, n);
      dfs(grid, i + 1, j, m, n);
      dfs(grid, i, j - 1, m, n);
      dfs(grid, i - 1, j, m, n);
    }
  }
};